src/history/hs_rec.c

```diff
@@ -557,22 +515,15 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
         if (oldest_upd->type == WT_UPDATE_TOMBSTONE && oldest_upd == first_non_ts_upd &&
           !F_ISSET(first_non_ts_upd, WT_UPDATE_CLEARED_HS)) {
             /* We can only delete history store entries that have timestamps. */
-            WT_ERR(__wt_hs_delete_key_from_ts(session, btree->id, key, 1));
+            WT_ERR(__wt_hs_reinsert_key_from_ts(session, btree->id, key, 1));
             WT_STAT_CONN_INCR(session, cache_hs_key_truncate_mix_ts);
-            clear_hs = false;
             F_SET(first_non_ts_upd, WT_UPDATE_CLEARED_HS);
-        } else
-            /*
-             * Clear the content with timestamps in the history store if we see updates without
-             * timestamps on the update chain.
-             *
-             * We don't need to clear the history store records if everything is still on the insert
-             * list and there are no updates moved to the history store by checkpoint or a failed
-             * eviction.
-             */
-            clear_hs = first_non_ts_upd != NULL &&
-              !F_ISSET(first_non_ts_upd, WT_UPDATE_CLEARED_HS) &&
-              (list->ins == NULL || ts_updates_in_hs);
+        } else if (first_non_ts_upd != NULL && !F_ISSET(first_non_ts_upd, WT_UPDATE_CLEARED_HS) &&
+          (list->ins == NULL || ts_updates_in_hs)) {
```

Comment:
@luke-pearson If you want to not clearing the history store if the mixed mode operation is on the data store, be careful to also handle this case:

U0 -> U20 -> U0. We still need to clear the history store even we have a U0 on data store.

Reply:
I think this is the scenario (U0 -> U20 -> U5):
U5 has ts 5 and is in the history store.
U20 is in the data store.
U0 gets added and written to the datastore, if we don't clear we then get U20 obsoleted and added to the datastore with ts 0 and then U0 gets moved the history store triggering the out of order deletion, as such U5 gets timestamp 0 and is moved after U20 and now we've got out of order updates. 

Is that your concern?

I think its best we just always clear the history store if we get a mixed mode operation and I don't implement my intended change. Do you agree?

Reply:
Yes, agree. Without the optimization is simpler.